### ISSUE TITLE
Add native support for the new M1/ARM based macs

### ISSUE THREAD

**Issue by cyperous** [sqlitebrowser/sqlitebrowser#2612]
Describe the new feature
--------------------------

Add native support for the new M1/ARM based macs.

Occasionally the application crashes, probably because of the emulation needed to run on M1/ARM based macs.

I search the issue list high and low and was actually quite surprised that no one else has posted this.

Additional info
---------------
Please answer these questions before submitting your feature request.

#### Is your feature request related to an issue? Please include the issue number.
NA

#### Does this feature exist in another product or project? Please provide a link.
Too many to list

#### Do you have a screenshot? Please add screenshots to help explain your idea.
NA

**Comment by justinclift** [comment]
It's a reasonable idea. :smile:

We'd either need to start using some external service to build the mac binaries, or perhaps get our own ARM based mac.  Given that we have a Patreon, it might be feasible for us to save up a few months to do that.

> Occasionally the application crashes, probably because of the emulation needed to run on M1/ARM based macs.

This bit (specifically) sounds like the emulator itself would be more stable with some bug fixes.  Personally, not being a macOS user though, I have no idea if that work is being done or not.  Hopefully it improves over time anway. :smile:

**Comment by scottfurry** [comment]
Save your pennies now - Mac Mini M1: $700 USD
https://www.apple.com/us/shop/buy-mac/mac-mini/apple-m1-chip-with-8-core-cpu-and-8-core-gpu-256gb

**Comment by justinclift** [comment]
Yeah, it'd be a 6 month saving thing. :wink:

**Comment by chrisjlocke** [comment]
It might be totally a different thing, but is https://www.appveyor.com/ any good for this sort of thing?  It seems to build from GitHub on different sources, etc.  It says for CI/CD, but naively assumed it compiled an application into different flavours.  Saw also https://octopus.com/features.
I have no idea what I'm on about though.... ;)

**Comment by justinclift** [comment]
@chrisjlocke Does AppVeyor definitely support ARM Macs?

Not seeing a mention of it, when quickly looking through what seems like the relevant pages:

* https://www.appveyor.com/docs/macos-images-software/
* https://www.appveyor.com/docs/build-environment/

Looking through the Octopus website, same question occurs... :wink:

**Comment by chrisjlocke** [comment]
Didn't look that far .... didn't even know if it might be useful for what we do ... just ... well ... er .... um ... 😣

**Comment by ngohungphuc** [comment]
Wonder when there will be version for m1 mac

**Comment by justinclift** [comment]
It'll be a few months away.  Currently our Patreon account is saving up funds for a new M1 (or whatever ARM) Mac.

**Comment by fishcharlie** [comment]
Any updates on this?

**Comment by elvey** [comment]
Thanks for working on this (and anyone helping fill the Patreon account or at least not whining. And if a discount helps, base M1 Mini is now $599 new, M1 Air is $850 refurb.)  Adding some keywords folks might be looking for to make this easier to find: 
 ARM-based macOS ARM64/Apple Silicon/M1 Max Tracking Issue.

**Comment by justinclift** [comment]
No worries at all. :smile:

Technically, we now have enough in our Patreon account to buy a new M1 based Mac Mini... it's just I'm personally slammed for time for the next few weeks (setting up infrastructure for a startup), and it's pretty likely I'd be the one looking after it.

You're quite welcome to proactively hassle me after the end of November (eg in 2-3 weeks), to make sure this doesn't drop off my radar. :smile:

**Comment by gennaios** [comment]
I think it's possible to build Universal binaries even on Intel. Qt officially added support with 6.2 though there are those who have got it working on 5.15.X. If older versions of Qt will work, I do not know. 

The bug that tracks the issue, with comments:

https://bugreports.qt.io/browse/QTBUG-85279

And a repo that seems useful if one wishes to remain on 5.x:

https://github.com/crystalidea/macdeployqt-universal

**Comment by justinclift** [comment]
As a data point, I just ordered the new M1 Mac Mini.  Hopefully we'll be able to build x86 binaries on it as well, and be able to retire the _very_ slow and ancient existing Mac Mini we're currently using for our nightly builds. :smile:

**Comment by gennaios** [comment]
Terrific. I think it is possible to build AS on Intel though without testing, you may not wish to pursue it. Btw, I was looking into building another Qt app on Apple Silicon, and someone mentioned that KDE maintains their own Qt 5.15.X with patches. You may wish to look into that if you're holding off the port to Qt 6.

https://community.kde.org/Qt5PatchCollection
https://invent.kde.org/qt/qt/qt5

**Comment by micsama** [comment]
> As a data point, I've just ordered the new M1 Mac Mini. Hopefully we'll be able to build x86 binaries on it as well, and be able to retire the _very_ slow and ancient existing Mac Mini we're currently using for our nightly builds. 😄

It's cool

**Comment by justinclift** [comment]
@gennaios Thanks, that looks like it'll be helpful.  The M1 Mac Mini turned up here today, and I was just now trying it out after work.

The hardware and software seems to work fairly well, though there are a few surprising other projects that don't have M1 builds available yet (Tor, and iGlance stand out).

Our DB4S builds have been using Qt 5.15.x so far, and we've not really investigated using Qt6.

Seeing as [The Qt Company](https://fuckyou.qt) have decided to make M1 macOS builds for Qt 15 only available to commercial users (:facepalm:), it sounds like we'll need to build our own Qt 5.15.x binaries and hope they work with ARM64.

Looking around online, it seems others have already hit this same problem and solved it for their own needs. eg:

* https://github.com/bobwolff68/fritzing-app/wiki/Building-Qt-5.15-for-Mac-M1
  * Mostly from: https://www.reddit.com/r/QtFramework/comments/ll58wg/how_to_build_qt_creator_for_macos_arm64_a_guide/

This looks to be potentially useful reference info while working on it:
* https://bugreports.qt.io/browse/QTBUG-85279
* https://github.com/nodejs/node/issues/40302

This is going to be interesting.  Will get to it in the next weeks, though it'll have to fit around other stuff.

**Comment by chrisjlocke** [comment]
That's odd - I can't get to that URL.  Hmm.  Must be my network.... 😉😊

![image](https://user-images.githubusercontent.com/7872218/145424692-df727344-f6b7-49c7-9634-0ed907819331.png)

**Comment by gennaios** [comment]
> The M1 Mac Mini turned up here today, and I was just now trying it out after work.

Terrific. :) I got mine over the weekend and it is a nice change.

>The hardware and software seems to work fairly well, though there are a few surprising other projects that don't have M1 builds available yet (Tor, and iGlance stand out).

Tor is native from homebrew. I haven't tried iGlance though I imagine if the homebrew-cask isn't, you may be able to compile it yourself, possibly with some fixes, or file a bug report.

>Our DB4S builds have been using Qt 5.15.x so far ..

From the previous links, I have been able to build a native version of Qt5 5.15.2 and QtWebEngine 5.15.7 (I believe). As the app I was trying to build has various dependencies that I've had trouble with and am still figuring out, I haven't completed such yet. but Qt5 is fine.

**Comment by gennaios** [comment]
I have since been able to build Qt 5.15.2 on Apple Silicon running Monterey and get the other app I was trying to build working. Since I had Qt already build, I gave this a go. It compiles fine, and seems to work ok.

In general to get Qt compiling on Monterey, besides various fixes listed above or mentioned elsewhere, MacPorts has some patches for Monterey as well as qmake, some of which I haven't seen elsewhere.

https://github.com/macports/macports-ports/tree/master/aqua/qt5/files

I believe all the patches I used came mainly from the Qt bug comments, as well as from the following link, and MacPorts, with the later possibly duplicating some of the previous.

https://github.com/crystalidea/qt-build-tools/tree/master/5.15.2

**Comment by justinclift** [comment]
Thanks.  A colleague pointed out that Homebrew likely has Qt compiled and working on M1 too.  So we could copy their formula and customise it for our needs as well.  In theory.

Timeline wise, I've just been stuck migrating servers to new infrastructure (bigger and better dedicated servers for DBHub.io).

My general plan for the next few weeks is:

1. Finish the main parts of our DBHub.io server migration
   * Likely today or tomorrow
2. Then get some business paperwork types of things done
   * Probably just after Xmas 
3. Then take a few days off
4. Then get some outstanding website updates completed
5. Then get this taken care of
   * Rough ETA of soon after new years day

**Comment by gennaios** [comment]
Whenever you get around to it is fine. Just wanted to let you know I was able to build.

A quick look at the Homebrew formula, it might not include the newer QtWebEngine 5.15.7 (at .8 last I checked) nor the necessary patches for such on Monterey, included in MacPorts. Seems this may not use it yet if you plan to use Qt for other projects, might as well.

**Comment by justinclift** [comment]
Been trying to get this work for a few hours today, and I'm calling it quits for tonight without success (yet).

So far, SQLite and SQLCipher are compiling fine, but when I try to get DB4S compiled and using (say) our self-compiled SQLite... it doesn't seem to.  eg:

```
[ 78%] Building CXX object CMakeFiles/sqlitebrowser.dir/src/sqlitetablemodel.cpp.o
/Users/jc/git_repos/sqlitebrowser/src/sqlitedb.cpp:2117:5: error: use of undeclared identifier 'sqlite3_enable_load_extension'
    sqlite3_enable_load_extension(_db, 1);
    ^
/Users/jc/git_repos/sqlitebrowser/src/sqlitedb.cpp:2121:18: error: use of undeclared identifier 'sqlite3_load_extension'
    int result = sqlite3_load_extension(_db, filePath.toUtf8(), nullptr, &error);
                 ^
/Users/jc/git_repos/sqlitebrowser/src/sqlitedb.cpp:2126:9: error: use of undeclared identifier 'sqlite3_enable_load_extension'
        sqlite3_enable_load_extension(_db, 0);
        ^
/Users/jc/git_repos/sqlitebrowser/src/sqlitedb.cpp:2144:5: error: use of undeclared identifier 'sqlite3_enable_load_extension'
    sqlite3_enable_load_extension(_db, Settings::getValue("extensions", "enable_load_extension").toBool());
    ^
4 errors generated.
make[2]: *** [CMakeFiles/sqlitebrowser.dir/src/sqlitedb.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/sqlitebrowser.dir/all] Error 2
make: *** [all] Error 2
```

Looking at the symbols in our compiled version of SQLite, they're present:

```
$ nm -gU /opt/homebrew/opt/sqlitefts5/lib/libsqlite3.0.dylib | grep -i extension
00000000000104c4 T _sqlite3_auto_extension
00000000000105f8 T _sqlite3_cancel_auto_extension
000000000001045c T _sqlite3_enable_load_extension
000000000000ffd8 T _sqlite3_load_extension
00000000000106c0 T _sqlite3_reset_auto_extension
```

So, it _seems_ like the problem (so far) is our CMake build process is refusing to use our SQLite, even though it's being provided via environment variables:

```
$ export CFLAGS="-I/opt/homebrew/opt/sqlitefts5/include/"
$ export CPPFLAGS="$CFLAGS"
$ export LDFLAGS="-L/opt/homebrew/opt/sqlitefts5/lib/"
$ cmake ..
$ make
```

Open to suggestions.  Hopefully it's just something simple (it's early am here, so not fully awake). :smile:

**Comment by scottfurry** [comment]
If I'm reading the situation and this [SO Answer](https://apple.stackexchange.com/a/410829) right, Apple Silicon has a different file system layout that Intel Silicon Macs.

We may have some paths or similar "assumptions" hard coded in our CMake files that could be complicating things. Look at 'target_link_directories' and 'link_directories' calls.

Found this cool [shell redirect](https://coder-question.com/cq-blog/198470) when using Homebrew. This might be useful.

Found a [link](https://dakota.sandia.gov/content/set-environment-linux-mac-os-x) that used a call to "export DYLD_LIBRARY_PATH=" to declare library paths.
Could this be the missing fix?

**Comment by gennaios** [comment]
I am less familiar with the dependencies and process so perhaps I cannot be of much help. After successfully building Qt 5.15.2 with QtWebEngine 5.15.7, all I did was follow the [build instructions](https://github.com/sqlitebrowser/sqlitebrowser/blob/master/BUILDING.md#building-an-app-bundle) and it just works. I tried again and it builds fine. Maybe you missed some of the needed Qt patches? There are many. Just a guess – I really do not know if that it is –, this is one:

https://github.com/macports/macports-ports/blob/master/aqua/qt5/files/patch-qmake-dont-hard-code-x86_64-as-the-architecture-when-using-qmake-installed.diff

**Comment by justinclift** [comment]
I'm finishing off a task for something else (not DB4S related) at the moment, which should be finished this week.  After that, my plan is to put time into DB4S stuff.  eg MacOS ARM building, next release, etc.

**Comment by 15921483570** [comment]
> I'm finishing off a task for something else (not DB4S related) at the moment, which should be finished this week. After that, my plan is to put time into DB4S stuff. eg MacOS ARM building, next release, etc.



> I'm finishing off a task for something else (not DB4S related) at the moment, which should be finished this week. After that, my plan is to put time into DB4S stuff. eg MacOS ARM building, next release, etc.

may i suggest i can use arm version in 1 month, like before 2022.8? 😎

**Comment by justinclift** [comment]
This last week I've been pretty much just chasing down and fixing one particular bug, that's nothing to do with ARM stuff though. :frowning_face:

**Comment by jtl001** [comment]
No problem, just keep us updated! 😄

**Comment by justinclift** [comment]
@gennaios With your comment above:

> Tor is native from homebrew.

How do you tell?  When I load Tor Browser, the "About" page just says it's "64-bit" with no mention of architecture. 🙄

Looking at the Tor project's bug tracker, the developers seem to have it successfully building on M1 now (for dev) but the releases don't include it:

&nbsp; &nbsp; https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/issues/40158

Note, only asking because I was looking at this while waiting for some stuff to compile on the M1 Mac mini today.

**Comment by gennaios** [comment]
I mean the tor daemon itself not the browser of which I’m not too familiar.

**Comment by tcurdt** [comment]
@justinclift `file` can tell you what binary it is. And I believe OpenSCAD is another candidate to have a look at for how they changed their build.

```
file /Applications/OpenSCAD.app/Contents/MacOS/OpenSCAD 
/Applications/OpenSCAD.app/Contents/MacOS/OpenSCAD: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64Mach-O 64-bit executable x86_64] [arm64]
/Applications/OpenSCAD.app/Contents/MacOS/OpenSCAD (for architecture x86_64):	Mach-O 64-bit executable x86_64
/Applications/OpenSCAD.app/Contents/MacOS/OpenSCAD (for architecture arm64):	Mach-O 64-bit executable arm64
```

**Comment by justinclift** [comment]
@gennaios No worries. :smile:

@tcurdt Oh, good thinking.  Thanks. :smile:

Using `file` on the latest Tor Browser (v11.5.1) download, it definitely only includes the x86_64 code:

```
$ file Tor\ Browser.app/Contents/MacOS/firefox 
Tor Browser.app/Contents/MacOS/firefox: Mach-O 64-bit executable x86_64
```

**Comment by mbalmer** [comment]
IIRC, any Mac, be it Intel or Apple Silicon based, can produce universal binaries. No need to buy an M1 mac. Id' love to see universal binaries for macOS.

**Comment by jtl001** [comment]
Hi, just checking in on progress for the Apple Silicon build. How's it going?

**Comment by gennaios** [comment]
It took a fair amount of effort to track down the Qt 5.15 patches for Apple Silicon. You may consider delaying til migration to Qt 6 if you think it’d be too much effort.

**Comment by justinclift** [comment]
Ahhh, interesting.  Lets see how it goes, after I've gotten the Intel macOS builds working again. :smile:

**Comment by justinclift** [comment]
Thanks @DingoBits, that was useful.  Got it compiling and running using that.  I *suspect* we achieve the same result by adjusting some paths in the CMakeLists.txt file, while I'm trying out now.

**Comment by justinclift** [comment]
Awesome, that worked.  I've just [added Qt5 and our custom SQLite paths to CMakeLists.txt](https://github.com/sqlitebrowser/sqlitebrowser/commit/c946fbd3d8df00cbbf45a469c5f0d1c33b8aa3dd), so they should now always be found by default if installed.

Next step, do some cleanup and testing, then get the other pieces of the packaging process working. :smile:

**Comment by justinclift** [comment]
This is a first tentative M1 build, using standard SQLite (not SQLCipher):

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/osx-prerelease/DB%20Browser%20for%20SQLite-arm64_20220924.dmg

It *seems* ok, except for the .app icon being busted.  Not sure what's wrong with that, but hopefully won't be too hard to fix.

Anyone up for testing it, and letting me know if it seems to work ok to you?

**Comment by jtl001** [comment]
Just tried to launch it and get a "can’t open the application because it may be damaged or incomplete" error

**Comment by tcurdt** [comment]
It works here. But I had to open it via right-click.
The `CFBundleIconFile` key in the `Info.plist` is empty.

**Comment by justinclift** [comment]
Thanks for testing! :smile:

The system it's building on is running up-to-date macOS Monterey (12.6).  I tested quickly in a freshly installed Monterey VM too (via [UTM](https://github.com/utmapp/UTM)).

Which macOS version(s) did you both test on?  Am wondering if it'll be that simple. :smile:

With my quick testing, I didn't need to right click it to launch it.  Might be some side effect of just rsync-ing it from the dev box?  Not sure (atm).

@tcurdt Ahhh, I'll update that for the next preliminary build.  Hopefully that's it. 🙏 

I've been staring at this stuff too much for one day.  I'll pick this up again tomorrow or Monday I guess. :smile:

**Comment by justinclift** [comment]
Thanks, that's really helpful.  It's clearly been a lot of time since I mucked around with this stuff. :smile:

**Comment by justinclift** [comment]
SQLCipher nightly builds now available as well:

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/latest/DB.Browser.for.SQLite-sqlcipher-arm64.dmg

(Also not notarized... yet :wink:)

**Comment by justinclift** [comment]
Gah, I broke the SQLCipher build script.  Fixing it...

**Comment by justinclift** [comment]
Fixed. :facepalm:

**Comment by justinclift** [comment]
If anyone's interested, this is the shell script (bash, not zsh) that's being used:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/blob/master/installer/macos/build_sqlitebrowser_nightly-arm64.sh

**Comment by justinclift** [comment]
Does anyone have a minute to try out this (un-notarized) build of DB4S 3.12.2 for arm64?

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/macos-prerelease/DB%20Browser%20for%20SQLite-arm64-3.12.2.dmg

In my limited testing just now, it seems like it's ok.  I _think_ it'll be workable to use as the base for adding to our Downloads page (eg after notarizing it).

**Comment by tcurdt** [comment]
@justinclift very(!) limited test - but it seems work OK.

**Comment by justinclift** [comment]
@tcurdt Thanks. :smile:

**Comment by justinclift** [comment]
This is a notarized version of our 3.12.2 release, build for arm64:

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/macos-prerelease/DB%20Browser%20for%20SQLite-arm64-3.12.2.dmg

I'll update the nightly build script to notarize the nightly builds as well.  Huge thanks to @DingoBits for the steps above. :smile:

**Comment by justinclift** [comment]
Our nightly build script for the arm64 (M1, M2, etc) binaries now automatically notarizes them.  I'll update the build script for our intel ones to do the same.

**Comment by justinclift** [comment]
Ugh, as the intel nightly build server is running an older macOS release (Catalina), it looks like the notarisation process is an older, much-more-pain-in-the-arse process.

I'll look at it some other time.

**Comment by justinclift** [comment]
Our nightly SQLCipher macOS builds have now been updated to use the latest point release of SQLCipher (4.5.2).

**Comment by justinclift** [comment]
The new arm64 build of DB4S 3.12.2 has been added to the download page on sqlitebrowser.org, and I just threw together a blog post about it + the arm64 nightly builds:

&nbsp; &nbsp;  https://sqlitebrowser.org/blog/apple-silicon-macos-package-available/

Closing this issue now, as I think it's all complete.  If I forgot something though, please someone speak up so it can be looked into. :smile:

**Comment by lucydodo** [comment]
@justinclift Awesome!

**Comment by mbalmer** [comment]
> Closed #2612 <https://github.com/sqlitebrowser/sqlitebrowser/issues/2612> as completed.
> 
Great!  Many thanks, that is much appreciated.
